### DISCUSSION TITLE
A command line Blog generator OR UI based predefined templates to select.

### DISCUSSION THREAD

**Discussion by salty-ivy** [wagtail/wagtail#10221] [Ideas]
As far as i have seen the most generic usecase for any CMS is the blog sites.  they help us create easy blog sites really fast.
this idea occured to me as i was creating multiple blog sites for my friends lately i was doing the same process over and over again...

### IDEA
i have seen what most of the application organizations do as they monitor the use-case by their clients as it becomes so common amongst people they sort build and a generic abstraction of that process and add it as ready to go feature.

this is what i propose for wagtail as well as wagtail is not only limited to blog site but blog sites is still grabs majority part of in usecase graph for any CMS.

we can have generator command that can build the basic boilerplate or ready to use structure for blogs.
something like `python3 manage.py createblog` that would do the following..
1. creating a blog app in code base.
2. populating `models.py` with generic blog structure like `BlogPage`  `BlogIndexPage` or `Article`  with common attributes like `title`
`body` `datatime` or may be `images`

3.creates a template and static files for it.
4. Equip the template with links to its static files CSS and JS.
5. may be adding a little  bit of styling already.

we can also get into more choosy part of this later on if implemented.

### Alternatives 
instead of command line, we could make this UI based. 
as individual package. what comes along with its data migrations, just like `modeladmin`, adding it to `installed apps`
and so on... 
we can provide some ready to use layouts  as well similar to `hashnode`   i.e `stacked` , `grid` etc.

**Comment by thibaudcolas** [comment]
@salty-ivy I think this is a good idea! Wagtail has a `wagtail start` command ([link to implementation](https://github.com/wagtail/wagtail/blob/main/wagtail/bin/wagtail.py#L388)) that uses Django’s `startproject` under the hood to generate projects based on a template.

We’ve discussed adding support for templates to this in the past. This would allow us to keep the "bootstrap a project" logic in Wagtail itself, and make it possible to support arbitrary templates. Some of those templates could be official ("simple blog"), and some could be from third parties.

---

There’s also been some work on doing this kind of project bootstrapping with a UI in the past – see for example https://wagtail.build/.

**Comment by thibaudcolas** [comment]
Just a note to say we’ve discussed this in today’s core team meeting, and are quite keen to invest in `wagtail start` to help with this. We think this will help with newcomers’ first experience, and also helping people produce higher-quality websites (as discussed above).

**Reply by thibaudcolas** [reply]
I’ve opened a new issue about fundamental support for templates in `wagtail start`: https://github.com/wagtail/wagtail/issues/10423.

**Reply by salty-ivy** [reply]
That's Great to hear! 🙌

**Comment by thibaudcolas** [comment]
👋 for anyone interested – we’re actively looking for feedback on people’s expectations with that kind of feature. For example:

- Which elements of similar tools should we replicate – Django `startproject` / `create-react-app` / `cookiecutter` / etc.
- Whether there are specific kinds of templates we should encourage, or try to build as "first party" templates. For example with specific front-end technologies, or perhaps ready-made StreamField blocks, or integration of specific packages on a Wagtail site

**Reply by pySilver** [reply]
@thibaudcolas I'd love to have the ability to inject my custom static resources into the admin static builder. Ideally, as a developer, I'd love to develop my custom widgets, stimulus controllers, styles, and other staff that extends Wagtail Admin and I'd love to be able to rebuild admin static files into new bundles that contain both original and custom code from external/my packages.

It may be useful for others to have such a feature for frontend, but me personally looking for admin integration only.

**Reply by thibaudcolas** [reply]
@pySilver could you expand on the type of integration you’d intend to do / share existing examples? This is generally something we want to support, but for now prioritise specific customisations as their cost is pretty high for us.

**Comment by bealdav** [comment]
Really nice idea thanks for all

I think, you should have a look here https://github.com/copier-org/copier
This tool could greatly help to maintain project skeleton between modified template and actively updated projects.
I think on this matter `copier` is really better than `cookiecutter`.

https://copier.readthedocs.io/en/latest

By example Copier allow to synchonize all 229 repo here https://github.com/OCA
Mainly of the repo have 7 branchs maintained https://github.com/OCA/oca-addons-repo-template#supported-use-cases 10.0 to 16.0
Here is the repo which help to maintain all the repos https://github.com/OCA/oca-addons-repo-template

My 2 cts

**Reply by thibaudcolas** [reply]
@bealdav could you expand on why copier is worth it? A big problem for us would be that we’d prefer to have this kind of "project bootstrapping" feature be built into the Wagtail package, so integrating something with as many dependencies as copier would be problematic.

**Reply by bealdav** [reply]
Oh you're right it could an advantage to have the feature built in wagtail.

- if you build it in wagtail, other django projects (or other python project) couldn't not use it.
- achieve a boostrap project tool is a thing which can be done inside wagtail for sure. Maintain it at low cost is another challenge
- achieve a project skeleton update (if the template is updated and/or the project is updated) is really really more difficult. Realize it inside wagtail prevent benefit others contributors to maintain it.

> Copier takes care of not overwriting existing files unless instructed to do so.

If you only want a bootstrap tool, don't use Copier. If you want update your project with the template updates, consider it

You may also have an optional installation with copier in pyproject.toml. Then normal installation is without it

I don't know if these informations answer to your question ?

my 2 cts

**Reply by thibaudcolas** [reply]
It does, thank you :) I’ll give copier a go and see what I think

**Comment by SebCorbin** [comment]
Another template idea: a very minimal template with Wagtail stripped down to the core (no contrib packages), no `home` package, just to use the Wagtail admin and editorial features on Django models.

That would also be a good starting point to monitor the coupling of some packages (e.g. `wagtail.sites` or `wagtail.search`).

**Comment by lb-** [comment]
All project templates should use a custom User, Images and Documents model. Or maybe opt out of this with a command arg. 

These are hard to change later.
